File: leetcode/303.range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
  this.prefixSum = nums.reduce(
    (sumArr, cur) => {
      const last = sumArr[sumArr.length - 1]
      sumArr.push(last + cur)
      return sumArr
    },
    [0]
  )
}

/**
 * @param {number} i
 * @param {number} j
 * @return {number}
 */
NumArray.prototype.sumRange = function(i, j) {
  return this.prefixSum[j + 1] - this.prefixSum[i]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(i,j)
 */

File: leetcode/303.range-sum-query-immutable.go
type NumArray struct {
	preSum []int
}

func Constructor(nums []int) NumArray {
	preSum := make([]int, len(nums)+1)
	preSum[0] = 0
	for i := 0; i < len(nums); i++ {
		preSum[i+1] = preSum[i] + nums[i]
	}
	return NumArray{
		preSum: preSum,
	}
}

func (this *NumArray) SumRange(i int, j int) int {
	return (*this).preSum[j+1] - (*this).preSum[i]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(i,j);
 */